glib-macros/src/closure.rs

```diff
@@ -1,193 +1,107 @@
-        } else {
-            None
-        };
-        if alias.is_none() {
-            if name.to_string() == "self" {
-                return Err(syn::Error::new_spanned(
-                    name,
-                    "Can't use `self` as variable name. Try storing it in a temporary variable or \
-                    rename it using `as`.",
-                ));
-            }
-            if name.to_string().contains('.') {
-                return Err(syn::Error::new(
-                    name.span(),
-                    format!(
-                        "`{}`: Field accesses are not allowed as is, you must rename it!",
-                        name
-                    ),
-                ));
-            }
-        }
-        Ok(Capture {
-            name,
-            alias,
-            kind,
-            start,
-        })
-    }
-}
+use crate::{
+    clone::{Capture, CaptureKind, DefaultReturn},
+    utils::crate_ident_new,
+};
 
 struct Closure {
     captures: Vec<Capture>,
     args: Vec<Ident>,
-    closure: syn::ExprClosure,
+    default_return: DefaultReturn,
+    closure: ExprClosure,
     constructor: &'static str,
 }
 
-impl syn::parse::Parse for Closure {
-    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+impl Parse for Closure {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
         let mut captures: Vec<Capture> = vec![];
-        if input.peek(Token![@]) {
-            loop {
-                let capture = input.parse::<Capture>()?;
-                if capture.kind == CaptureKind::Watch {
+        let mut default_return: DefaultReturn = DefaultReturn::default();
+
+        loop {
+            // There must either be one or no attributes here. Multiple attributes are not
+            // supported.
+            //
+            // If this is a capture attribute, it must be followed by an identifier.
+            // If this is a default return attribute, it might be followed by a closure. After the
+            // default return attribute there must not be any further attributes.
+            //
+            // If this is not an attribute then it is a closure, async closure or async block which
+            // is handled outside the loop
+            let attrs = input.call(Attribute::parse_outer)?;
+            let Some(attr) = attrs.first() else {
+                break;
+            };
+            let attr_name = attr.path();
+            if let Some(capture_kind) = attr_name
+                .get_ident()
+                .and_then(|i| CaptureKind::try_from(i).ok())
+            {
+                if capture_kind == CaptureKind::Watch {
                     if let Some(existing) = captures.iter().find(|c| c.kind == CaptureKind::Watch) {
                         return Err(syn::Error::new(
                             existing.start,
-                            "Only one `@watch` capture is allowed per closure",
+                            "Only one watch capture is allowed per closure",
```

Comment:
Didn't see any test for this error. Please add one.

Reply:
Also, please put `watch` between backticks.

Reply:
There are no tests for `closure!` error messages at all

Reply:
I.e. I'll add those